### ISSUE TITLE
False positive "Cannot call method on possible iterable" for intersection types

### ISSUE THREAD

**Issue by VincentLanglet** [vimeo/psalm#6340]
I might be wrong but to me if a method existe on Foo, it exists on Foo&iterable.
(I agree it might not exist on Foo|iterable)

https://psalm.dev/r/9444a0ce63

**Comment by TysonAndre** [comment]
It seems like a possible way to handle this would be to convert `iterable<...>` to `\Traversable<...>` if the other types in the intersection type must be an object, when creating intersection types?

src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/AtomicMethodCallAnalyzer.php 

It seems like
1. Psalm treats iterable on its own as possibly invalid. That makes sense
2. The way that individual types in the intersection type are merged into the type depends on the order in some way. Though I assume this is still mostly correct, since the `analyze` is acting on a copy of the `$result` and appending to arrays or overwriting the values

(`@param iterable<string, mixed>&Foo $foo` results in InvalidMethodCall rather than PossiblyInvalidMethodCall)

```php
        foreach ($intersection_types as $intersection_type) {
            $intersection_result = clone $result;

            /** @var ?Type\Union */
            $intersection_result->return_type = null;

            self::analyze(
                $statements_analyzer,
                $stmt,
                $codebase,
                $context,
                $lhs_type,
                $intersection_type,
                $lhs_type_part,
                true,
                $lhs_var_id,
                $intersection_result
            );

            $result->returns_by_ref = $intersection_result->returns_by_ref;
            $result->has_mock = $intersection_result->has_mock;
            $result->has_valid_method_call_type = $intersection_result->has_valid_method_call_type;
            $result->has_mixed_method_call = $intersection_result->has_mixed_method_call;
            $result->invalid_method_call_types = $intersection_result->invalid_method_call_types;
            $result->check_visibility = $intersection_result->check_visibility;
            $result->too_many_arguments = $intersection_result->too_many_arguments;
```
